File: go/pdfcontext/ports.go
package pdfcontext

import "pdftoolkit/sharedkernel"

// Ports (interfaces). Adapters in infra_*.go implement these — mirrors the Python
// ABCs and Rust traits of the same names.

type Splitter interface {
	Split(source sharedkernel.MediaFile, specs []SplitSpec, outDir string) ([]sharedkernel.MediaFile, error)
}

type PdfInspector interface {
	Inspect(source sharedkernel.MediaFile) (DocumentCensus, error)
}

type PdfCodec interface {
	ExtractImage(source sharedkernel.MediaFile, xref int) ([]byte, string, error)
	StructuralOptimize(source sharedkernel.MediaFile, out string) (sharedkernel.MediaFile, error)
}

// Compressor is a whole-document compressor. GS, pdfcpu-native, etc. all fit here.
type Compressor interface {
	Name() string
	Compress(source sharedkernel.MediaFile, target CompressionTarget, out string) (CompressionResult, error)
}

File: rust/src/pdf_context/ports.rs
//! pdf context — ports (traits). Adapters in infra_*.rs implement these.

use std::path::Path;

use super::domain::{CompressionResult, CompressionTarget, DocumentCensus, SplitSpec};
use crate::shared_kernel::MediaFile;

pub trait Splitter {
    fn split(
        &self,
        source: &MediaFile,
        specs: &[SplitSpec],
        out_dir: &Path,
    ) -> Result<Vec<MediaFile>, String>;
}

pub trait PdfInspector {
    fn inspect(&self, source: &MediaFile) -> Result<DocumentCensus, String>;
}

/// A whole-document compressor. GS, lopdf-native, etc. all fit here.
pub trait Compressor {
    fn name(&self) -> &str;
    fn compress(
        &self,
        source: &MediaFile,
        target: &CompressionTarget,
        out: &Path,
    ) -> Result<CompressionResult, String>;
}
